### PR TITLE
Implement ranges::sort

### PR DESCRIPTION
**Pull request by CaseyCarter**
Partially addresses #39.

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
@AlexGuteniev @BillyONeal This PR's test run failed:
```cpp
Command: "D:\build\x86\tests\std\tests\P1135R6_atomic_wait_vista\00\P1135R6_atomic_wait_vista.exe"
Exit Code: 3
Standard Error:
--
Assertion failed: strcmp(seq, "123456") == 0, file C:/agent/_work/1/s/tests/std/include\test_atomic_wait.hpp, line 55
--
```

**Comment by AlexGuteniev** [comment]
@StephanTLavavej , I'm afraid this is due to timing assumption. Unfortunately, atomic wait test has to use timing assumption.

File: stl/inc/algorithm [hunk 1]
```diff
@@ -6330,6 +6330,22 @@ _CONSTEXPR20 void make_heap(_RanIt _First, _RanIt _Last) { // make [_First, _Las
 #ifdef __cpp_lib_concepts
 namespace ranges {
     // VARIABLE ranges::make_heap
+    // clang-format off
+    template <random_access_iterator _It, class _Pr, class _Pj>
+        requires sortable<_It, _Pr, _Pj>
+    constexpr void _Make_heap_common(_It _First, _It _Last, _Pr _Pred, _Pj _Proj) {
+        // clang-format on
+        // make nontrivial [_First, _Last) into a heap
```
Comment (CaseyCarter) [inline stl/inc/algorithm:6338]: What does "nontrivial" mean here?

File: stl/inc/algorithm [hunk 2]
```diff
@@ -7594,9 +7599,78 @@ namespace ranges {
...
+
+    // VARIABLE ranges::sort
+    class _Sort_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <random_access_iterator _It, sentinel_for<_It> _Se, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<_It, _Pr, _Pj>
+        _NODISCARD constexpr _It operator()(_It _First, _Se _Last, _Pr _Pred = {}, _Pj _Proj = {}) const {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:7611]: I believe that `sort`'s public function call operators shouldn't be marked `_NODISCARD` - 99% of callers will want to discard here.

File: stl/inc/algorithm (hunk 1 above)
Comment (StephanTLavavej) [reply stl/inc/algorithm:6338]: No idea, I think we should fix up the pre-existing occurrences.

File: stl/inc/algorithm [hunk 3]
```diff
@@ -7594,9 +7599,78 @@ namespace ranges {
...
+            return _First;
+        }
+
+        // clang-format off
+        template <random_access_range _Rng, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<iterator_t<_Rng>, _Pr, _Pj>
+        _NODISCARD constexpr borrowed_iterator_t<_Rng> operator()(_Rng&& _Range, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            // clang-format on
+            auto _ULast       = _Get_final_iterator_unwrapped(_Range);
+            auto _UFirst      = _Ubegin(_Range);
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:7628]: Style: It's a little strange to see `_ULast` defined before `_UFirst`.

File: stl/inc/algorithm [hunk 4]
```diff
@@ -7594,9 +7599,78 @@ namespace ranges {
...
+            const auto _Count = _ULast - _UFirst;
+            _Sort_common(_STD move(_UFirst), _ULast, _Count, _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return _Rewrap_iterator(_Range, _STD move(_ULast));
+        }
+
+    private:
+        template <class _It, class _Pr, class _Pj>
+        _NODISCARD static constexpr void _Sort_common(
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:7636]: `_NODISCARD` shouldn't appear with `void`.

File: stl/inc/algorithm [hunk 5]
```diff
@@ -7594,9 +7599,78 @@ namespace ranges {
...
+
+                _Ideal = (_Ideal >> 1) + (_Ideal >> 2); // allow 1.5 log2(N) divisions
+
+                if (_Mid_first - _First < _Last - _Mid_last) { // loop on second half
+                    _Sort_common(_First, _Mid_first, _Ideal, _Pred, _Proj);
+                    _First = _Mid_last;
+                } else { // loop on first half
+                    _Sort_common(_Mid_last, _Last, _Ideal, _Pred, _Proj);
+                    _Last = _Mid_first;
+                }
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:7666]: As `[_Mid_first, _Mid_last]` are about to go out of scope, should their final uses here be moved-from?

File: stl/inc/algorithm [hunk 6]
```diff
@@ -7594,9 +7599,78 @@ namespace ranges {
...
+
+                if (_Ideal <= 0) { // heap sort if too many divisions
+                    _RANGES _Make_heap_common(_First, _Last, _Pred, _Proj);
+                    _RANGES _Sort_heap_common(_STD move(_First), _STD move(_Last), _Pred, _Proj);
+                    return;
+                }
+
+                // divide and conquer by quicksort
+                auto [_Mid_first, _Mid_last] =
+                    _RANGES _Partition_by_median_guess_unchecked(_First, _Last, _Pred, _Proj);
```
Comment (miscco) [inline stl/inc/algorithm:7656]: Should this also be suffixed  with `common`

File: tests/std/tests/P0896R4_ranges_alg_sort/test.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+using namespace std;
+using P = pair<int, int>;
+
+// Validate dangling story
+STATIC_ASSERT(same_as<decltype(ranges::sort(borrowed<false>{})), ranges::dangling>);
+STATIC_ASSERT(same_as<decltype(ranges::sort(borrowed<true>{})), int*>);
+
+struct instantiator {
+    static constexpr array input = {P{-1200257975, 0}, P{-1260655766, 1}, P{-1298559576, 2}, P{-1459960308, 3},
```
Comment (miscco) [inline tests/std/tests/P0896R4_ranges_alg_sort/test.cpp:21]: Do we need array here or can we increase throughput by using a plain C-style array

File: stl/inc/algorithm (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:7628]: This is an attempt to help the compiler avoid spilling `_UFirst` into memory while evaluating `_Get_final_iterator_unwrapped(_Range)`, which is a bit silly given that I've never so much as looked at codegen for any specialization of this template. I'll reorder these.

File: stl/inc/algorithm (hunk 5 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:7666]: Yes, that is more consistent with the policies I've been using for discretionary iterator moves.

File: tests/std/tests/P0896R4_ranges_alg_sort/test.cpp (hunk 7 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_sort/test.cpp:21]: Replacing this with a C array - and replacing the two copy constructions in the body of the test with calls to `ranges::copy` - is a wash on my machine. Fastest of ten runs of all 17 configurations is 5.52s (`std::array`) vs. 5.55s (C array).

The `array` initialization is uglier - the `P`s are necessary for CTAD to work - but the copy construction syntax at the use sites is superior. so readability is a bit of a wash as well. I see no reason to prefer either approach so I'll stick with the status quo.

File: stl/inc/algorithm (hunk 6 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:7656]: I've been using `_common` for functions with the same iterator and sentinel type where there's a reasonable expectation that we might want to add an `_unchecked` variation that accepts differing iterator and sentinel types, and simply using `_unchecked` when I believe no such expectation exists. I now think that policy wasn't well thought out, and it makes more sense to consistently use `_common` consistently for Ranges algorithms that don't unwrap _and_ don't accept differing iterator and sentinel types.

So yes.

**Comment by StephanTLavavej** [comment]
Thanks for getting this algorithm *sorted* out! :joy_cat:

**Pull request merged**